src/extension/messaging.ts

```diff
@@ -0,0 +1,120 @@
+import { join } from 'node:path'
+
+import { Disposable, ExtensionContext, Uri, window, workspace } from 'vscode'
+import { TelemetryReporter } from 'vscode-telemetry'
+
+import { EXTENSION_NAME, TELEMETRY_EVENTS } from '../constants'
+
+import { fileOrDirectoryExists, getDefaultWorkspace, isMultiRootWorkspace } from './utils'
+
+export abstract class DisplayableMessage {
+    abstract dispose(): void
+    abstract display(): void
+    constructor(protected readonly context: ExtensionContext, readonly stateSettings?: Record<TELEMETRY_EVENTS, any>) {
+        this.context = context
+        this.stateSettings = stateSettings
+    }
+}
+
+export class MessagingBuilder implements Disposable {
+    /**
+     * Represents a collection of DisplayableMessages, useful to display window messages.
+     * @param messages Specify the displayable messages to show
+     */
+    constructor(private messages: DisplayableMessage[]) {
+        this.messages = messages
+    }
+
+    dispose() {
+        this.messages.forEach(d => d.dispose())
+    }
+
+    /**
+     * Display registered DisplayableMessages
+     */
+    activate() {
+        for (const message of this.messages) {
+            message.display()
+        }
+    }
+}
+
+
+export class RecommendExtensionMessage extends DisplayableMessage implements Disposable {
+    async display(): Promise<void> {
+        try {
+            const skipPrompSettings = this.stateSettings && this.stateSettings[TELEMETRY_EVENTS.RecommendExtension]
+            let promptUser = skipPrompSettings ||
+                this.context.globalState.get(TELEMETRY_EVENTS.RecommendExtension, true)
```

Comment:
@degrammer, I believe there's a workspaceState or something like that? Should be record the "skipping" globally? Meaning it will never ever be shown again? Irrespective of the repo one's working on 🤔. I am thinking out loud here.

Reply:
@sourishkrout  global state works across repositories. Using workspaceState will do the opposite of what you want.

Reply:
Well, what I want - I think - is that if I dismissed the recommendation for a particular repo once that it won't apply to all future repos, I am opening. Perhaps we need a "No, thanks" and a "Never ask again" button?